### PR TITLE
search / saving / '0' column / redrawing after editing bytes / 2 minor bugs

### PR DESCRIPTION
**Pull request by rankinbc**
-Added search. You can search for all addresses containing a string of bytes or a specific value and data type. You can search all addresses or just search for aligned values.

-Added saving. Save changes to file with (Ctrl+S).

-Added a '0' column to the default columns so that the hex values align better with the values' addresses.

-Fixed bug where program would crash if ASCII values area was clicked before loading a file.

-When the scrollbar position was not 0, editing a byte would change the
Data[] but the byte was not being redrawn. Added an invalidate() at the
end of the RedrawCurrentByte method to redraw the updated bytes.

--Sometimes the buffer would have a single character stuck in it making
it frustrating to write the correct value. The buffer is now emptied
when a new byte is selected.

### CODE REVIEW

**Comment by rankinbc** [comment]
The editing functionality appearing to be broken was not caused by my earlier commits. The data array was being edited, but the rectangles were not being redrawn correctly. If you scrolled up and then down once, the updated data would be shown. 

The "Invalidate(rect)" for the hex and ASCII section was only updating the rectangles if the scrollbar was at position 0. Invalidating the whole "hex21" fixed the problem. There might be a better way than redrawing the whole section, but it doesn't noticeably slow the program down. The only effect is that it flashes similar to how it does when you scroll down.

**Comment by r1sc** [comment]
Sorry I haven't checked this earlier. I'll look at it tonight after work.

File: hex/Hex2.cs [hunk 1]
```diff
@@ -163,12 +173,16 @@ class Hex2 : UserControl {
...
-            Invalidate(rect);
+            Invalidate(rect);       
 
             charWidth = (int)Font.Size;
             asciiStart += charWidth;
             rect = new Rectangle(asciiStart + col * charWidth, row * 24, charWidth, 24);
             Invalidate(rect);
+
+            //redraws the whole section, not sure if there's a better way.
+            Invalidate();
```
Comment (r1sc) [inline hex/Hex2.cs]: This is not a good solution. We need to recalculate the rect with respect to the window scrolling position.

File: hex/Hex2.resx [hunk 2]
```diff
@@ -0,0 +1,120 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (r1sc) [inline hex/Hex2.resx:1]: What's the point of this file?

File: hex/FormTest2.resx [hunk 3]
```diff
@@ -120,6 +120,15 @@
   <metadata name="statusStrip1.TrayLocation" type="System.Drawing.Point, System.Drawing, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a">
     <value>17, 17</value>
   </metadata>
+  <data name="hex21.QueryHitAddresses" mimetype="application/x-microsoft.net.object.binary.base64">
```
Comment (r1sc) [inline hex/FormTest2.resx:123]: What's this?

File: hex/Hex2.cs [hunk 4]
```diff
@@ -242,7 +257,7 @@ class Hex2 : UserControl {
 
                 rect.Width = charWidth;
                 for (int i = 0; i < numColumns; i++) {
-                    var num = ((i % 15) + 1);
+                    var num = ((i % 16));
```
Comment (r1sc) [inline hex/Hex2.cs:258]: I don't agree with 0 being the first column. I still think it should be 1.

File: hex/Hex2.cs [hunk 5]
```diff
@@ -233,7 +248,7 @@ class Hex2 : UserControl {
                 e.Graphics.FillRectangle(Brushes.LightGray, new Rectangle(0, 0, ClientRectangle.Width, 24));
 
                 for (int i = 0; i < numColumns; i++) {
-                    var num = (i + 1);
```
Comment (r1sc) [inline hex/Hex2.cs:236]: I don't agree with 0 being the first column. I still think it should be 1.

File: hex/Hex2.cs [hunk 6]
```diff
@@ -163,12 +173,16 @@ class Hex2 : UserControl {
             var charWidth = (int)(Font.Size * 2);
             var asciiStart = charWidth * numColumns;
             rect = new Rectangle(col * charWidth, row * 24, charWidth, 24);
-            Invalidate(rect);
+            Invalidate(rect);
```
Comment (r1sc) [inline hex/Hex2.cs]: ?

File: hex/Hex2.cs [hunk 7]
```diff
@@ -38,12 +40,13 @@ class Hex2 : UserControl {
...
-            Controls.Add(_scrollBar);
+            Controls.Add(_scrollBar); 
+            QueryHitAddresses = new List<Int32>();
         }
 
         void _scrollBar_Scroll(object sender, ScrollEventArgs e) {
-            Invalidate();
-            UpdateCaret(CurrentPos);
+            Invalidate();       
+            UpdateCaret(CurrentPos);
```
Comment (r1sc) [inline hex/Hex2.cs:49]: ?

File: hex/Hex2.cs (hunk 1 above)
Comment (rankinbc) [inline hex/Hex2.cs]: Yeah this was more of a quick fix so that the editing wouldn't break after the scrollbar is moved. This bug is in the original repository though, it's not caused by my commits.

File: hex/Hex2.resx (hunk 2 above)
Comment (rankinbc) [inline hex/Hex2.resx:1]: Sorry, this was generated by Visual Studio. I'm still learning Github and haven't got down what files to remove from commits yet. I tried to remove all the unnecessary changes from the commit but I must have missed some. This change can be removed without affecting the program.

File: hex/FormTest2.resx (hunk 3 above)
Comment (rankinbc) [inline hex/FormTest2.resx:123]: QueryHitAddresses is the list of addresses which match the search criteria when you search for something. Without this code, the list will always return null.

File: hex/Hex2.cs (hunk 4 above)
Comment (rankinbc) [inline hex/Hex2.cs:258]: Ok that's fine. Most hex editors I've seen have 16 columns so that the data will align with the actual memory address. 
hex

If you don't mind me asking, why do you prefer it start at 1?

File: hex/Hex2.cs (hunk 6 above)
Comment (rankinbc) [inline hex/Hex2.cs]: Oops sorry about that, didn't know that extra whitespace would be counted as a change.
